api/v1alpha3/cluster_types.go

```diff
@@ -183,3 +187,56 @@ type ClusterList struct {
 func init() {
 	SchemeBuilder.Register(&Cluster{}, &ClusterList{})
 }
+
+// FailureDomains is a slice of FailureDomain objects.
+type FailureDomains []FailureDomain
+
+// Add adds the FailureDomain to the list of FailureDomains, it returns an error if the id is a duplicate.
+func (in *FailureDomains) Add(domain FailureDomain) error {
```

Comment:
I still think we should likely also update the cluster validation webhook to verify uniqueness as well, but this `Add` method will definitely help downstream consumers to implement this safely.

Should we also have an `Update` method to make it easier to update the value for a given failure domain by id?

Reply:
SGTM